index.js

```diff
@@ -38,11 +38,15 @@ export function trimLines(value) {
  */
 function trimLine(value, start, end) {
   if (!start) {
-    value = value.replace(/^[ \t]+/, '')
+    const regex = /^[ \t]+/
+    value = value.replace(regex, '')
   }
 
   if (!end) {
-    value = value.replace(/[ \t]+$/, '')
+    const regex = /[ \t]/
+    let i = value.length
+    while (regex.test(value.charAt(--i)));
+    return value.slice(0, i + 1)
   }
 
   return value
```

Comment:
At this point, it seems like a bad idea to use regexes for this. At least, it’s hard to read. Can you check if something like this works?

```js
let start = 0
let end = value.length - 1

while (start < end) {
    let code = value.charAt(start);
    if (code === 9 || code === 32) {
        start++
    } else {
        break;
    }
}

while (end > start) {
  let code = value.charAt(start);
  if (code === 9 || code === 32) {
    end--;
  } else {
    break;
  }
}

return value.slice(start, end);
```

(formatted to match project)
(perhaps good to check for some lines that just include whitespace, and line that include nothing at all)

Reply:
I understand your opinion, but I disagree. Can you change it?
Regexes are generally slow. Character codes, especially with parsing projects such as all of unified, are common and searchable.

Reply:
I propose the follow as the function. Please still consider it pseudocode, however, I did test it and it seems to work

```js
/**
 * @param {string} value
 * @param {boolean} start
 * @param {boolean} end
 * @returns {string}
 */
function trimLine(value, start, end) {
    let startIndex = 0
    let endIndex = value.length

    if (!start) {
        let code = value.charCodeAt(startIndex)

        while (code === 9 || code === 32) {
            startIndex++
            code = value.charCodeAt(startIndex)
        }
    }

    if (!end) {
        let code = value.charCodeAt(endIndex - 1)

        while (code === 9 || code === 32) {
            endIndex--
            code = value.charCodeAt(endIndex - 1)
        }
    }

    return endIndex > startIndex ? value.slice(startIndex, endIndex) : ''
}
```

This:

- Implements faster trimming for `!start` as well, meaning that regexes are no longer needed
- Uses `charCodeAt` which is faster that `charAt`, and no longer needs small strings. I understand that these codes might be new to you, and hence you do not prefer them, but I consider them common enough in parsing, in the 100s of projects I am maintaining, that I strongly prefer them.
- Changes `value` only once, without reassigning it, or even not at all for empty lines. Reassigning a parameter is slow, because JavaScript “links” `arguments[0]` and `value` together. Not slicing at all for empty lines is likely also fast in edge cases of large blank lines.

All reasons why this *should* be faster.

Reply:
I suggest using constants for the character codes to make it more readable:

```js
const FOO = 9

//...


while (code === FOO || code === BAR) {
```

And I agree with not using regex, both for readability, performance, and also to avoid the risk of ReDoS.

Reply:
> I suggest using constants for the character codes to make it more readable:
> 
> ```js
> const FOO = 9
> 
> //...
> 
> 
> while (code === FOO || code === BAR) {
> ```
> 
> And I agree with not using regex, both for readability, performance, and also to avoid the risk of ReDoS.

Just saw this, I think it resolves my concern about readability without compromising performance. I'll make the constants now.